README
CapGun
======

Introduction goes here.



Example
=======

Example goes here.

Issues
======
Totally untested and probably won't work on non *nix platform -- patches welcome!


Copyright (c) 2008 Relevance, Inc., released under the MIT license


CHANGELOG
v0.2.4 Clean up and a little more safety around previous revision

v0.2.3  Fix the case where there is no previous revision so it doesn't blow up

v0.2.2  SVN support courtesy Olivier Brisse

v0.2.1  Documentation update so that it works with ActionMailer 2.3+

v0.2.0  Git changes in the email, major refactoring

v0.0.9  Ruby 1.8.7 support via a small change to the Net::SMTP SSL hack

v0.0.8. Fix failing spec on runcoderun that depended on timezone; add redgreen to specs

v0.0.7. Pick up changelog and gemspec

v0.0.6. Fix subject line if there *is* a rails_env defined - this was a bug introduced with the 0.0.5 'fix'.  Awesome.

v0.0.5. Fix subject line iv there is no rails_env defined

v0.0.4. Add dev dependency on echoe; update manifest

v0.0.3. Don't try to convert the RAILS_ENV if its not set to better support default capistrano deployments

v0.0.2. Converts the release times to the current time zone; improved docs.

v0.0.1. Initial release.  Rock n roll mcdonalds.


spec/cap_bot_spec.rb
ENV["RAILS_ENV"] = "test"
require 'rubygems'
require 'test/unit'
require 'test/spec'
require 'mocha'
require 'net/smtp'
require File.join(File.dirname(__FILE__), *%w[.. lib cap_bot])

describe "CapGun" do
  it "uses action mailer hack" do
    Net::SMTP.new('').respond_to?(:starttls, true).should == true
  end
end

describe "CapGun" do
  include CapGun::Helper
  
  it "parses datetime from release path" do
    time_from_release("/data/foo/releases/20080402152141", "PDT").should == "April 2nd, 2008 3:21 PM PDT"
  end
end

spec/cap_gun_spec.rb
ENV["RAILS_ENV"] = "test"
require 'rubygems'
require 'test/unit'
require 'test/spec'
require 'mocha'
require 'net/smtp'
require 'redgreen' unless Object.const_defined?("TextMate")

require File.join(File.dirname(__FILE__), *%w[.. lib cap_gun])

describe "CapGun" do
  it "uses action mailer hack" do
    Net::SMTP.new('').respond_to?(:starttls, true).should == true
  end
end

describe "CapGun" do

  describe "mail settings" do
    
    it "raises if we don't have settings" do
      capistrano = stub_everything
      lambda { CapGun::Mailer.load_mailer_config(capistrano) }.should.raise(ArgumentError).message.should == "You must define ActionMailer settings in 'cap_gun_action_mailer_config'"
    end
  
    it "gets action mailer config from capistrano" do
      capistrano = stub(:cap_gun_action_mailer_config => {:account => "foo@gmail.com", :password => "password"}, :exists? => true)
      capistrano.stubs(:[]).returns({:recipients => 'foo'})
      CapGun::Mailer.load_mailer_config(capistrano)
      ActionMailer::Base.smtp_settings.should == {:account => "foo@gmail.com", :password => "password"}
    end
    
    it "raises if don't have a cap gun email envelope" do
      capistrano = stub_everything(:cap_gun_action_mailer_config => {}, :exists? => false)
      lambda { CapGun::Mailer.load_mailer_config capistrano }.should.raise(ArgumentError)
    end
    
    it "raises if we don't have at least one recipient" do
      capistrano = stub_everything(:cap_gun_action_mailer_config => {}, :cap_gun_email_envelope => {})
      lambda { CapGun::Mailer.load_mailer_config capistrano }.should.raise(ArgumentError)
      capistrano = stub_everything(:cap_gun_action_mailer_config => {}, :cap_gun_email_envelope => {:recipients => []})
      lambda { CapGun::Mailer.load_mailer_config capistrano }.should.raise(ArgumentError)
    end
    
  end
  
  describe "Mailer" do
    
    it "calls Net::SMTP to send the mail correctly (we test this because SMTP internals changed between 1.8.6 and newer versions of Ruby)" do
      ActionMailer::Base.smtp_settings = {
        :address => "smtp.gmail.com",
        :port => 587,
        :domain => "foo.com",
        :authentication => :plain,
        :user_name => "username",
        :password => "password"
      }
    
      capistrano = { :current_release => "/data/foo", :previous_release => "/data/foo", :cap_gun_email_envelope => {:recipients => ["joe@example.com"]} }        
      smtp = Net::SMTP.new('gmail.com', 25)
      Net::SMTP.expects(:new).returns(smtp)
      smtp.expects(:start)
      CapGun::Mailer.deliver_deployment_notification capistrano
    end
    
  end
  
end

lib/cap_bot.rb
require 'active_support'
require 'action_mailer'
require File.join(File.dirname(__FILE__), *%w[.. vendor action_mailer_tls lib smtp_tls])

# Tell everyone about your releases!
# 
# You must set up your mail settings in config/capinator_config.yml.  We include
# the ActionMailer hack to play nice with gmail, so thats a super easy way 
# to do this without setting up your own MTA.
#
# Example:
#
# Want to just shoot everyone an email about the latest status?
#
#   cap capinator:email
# 
# Include comments?
#
#   cap -s comment="hi mom" capinator:email
#
# Enable emails after every deploy by adding this to your deploy.rb:
# 
#   after "deploy", "capinator:email"
#
# Then when you deploy, you can optionally include comments:
#   cap -s comment="fix for bug #303" deploy
#
module Capinator

  module Helper
    
    def load_mailer_config
      mailer_config = File.open("#{rails_root}/config/capinator_config.yml")
      mailer_options = YAML.load(mailer_config)
      ActionMailer::Base.smtp_settings = mailer_options
    end
  
    # Capistrano doesn't set this for some reason
    def rails_root
      Object.const_set("RAILS_ROOT", File.expand_path(File.join(File.dirname(__FILE__), *%w[.. .. .. ..]))) unless Object.const_defined?("RAILS_ROOT")
      RAILS_ROOT
    end
  
    def current_user
      `id -un`.strip
    end
  
    def time_from_release(path, timezone)
      timestamp = path[(path.rindex("/") + 1)..-1]
      datetime = DateTime.parse(timestamp)
      datetime.strftime("%B #{datetime.day.ordinalize}, %Y %l:%M %p #{timezone}").gsub(/\s+/, ' ').strip
    end
    
    extend self
  end
  
    class Mailer < ActionMailer::Base
      include Capinator::Helper
      DEFAULT_SENDER = %("Capinator" <capinator@example.com>)

      @@email_prefix = "[DEPLOY] "
      cattr_accessor :email_prefix
      
      def deployment_notification(capistrano)
        options = capistrano[:capinator_options]
        
        content_type "text/plain"
        subject "#{email_prefix} #{capistrano[:application]} deployed to #{capistrano[:rails_env]}"

        recipients options[:recipients]
        from       options[:sender_address] || DEFAULT_SENDER

        body       <<-EOL
#{capistrano[:application]} was deployed to #{capistrano[:rails_env]} by #{current_user} at #{time_from_release(capistrano[:current_release], capistrano[:timezone])}.

Comment: #{capistrano[:comment] || "[none given]"}

Nerd details
============
Release: #{capistrano[:current_release]}
Release Time: #{time_from_release(capistrano[:current_release], capistrano[:timezone])}
Release Revision: #{capistrano[:current_revision]}

Previous Release: #{capistrano[:previous_release]}
Previous Release Time: #{time_from_release(capistrano[:previous_release], capistrano[:timezone])}
Previous Release Revision: #{capistrano[:previous_revision]}

Repository: #{capistrano[:repository]}
Deploy path: #{capistrano[:deploy_to]}
EOL
      end

    end
  

end

if Object.const_defined?("Capistrano")

  Capistrano::Configuration.instance(:must_exist).load do

    namespace :capinator do
      desc "Get the time zone from the server"
      task :get_timezone, :roles => :app do
        set "timezone", capture('date "+%Z"').strip
      end
      
      desc "Send notification via email"
      task :email do
        Capinator::Helper.load_mailer_config
        Capinator::Mailer.deliver_deployment_notification(self)
      end

    end
    
    before "capinator:email", "capinator:get_timezone"
    
  end
end

lib/cap_gun.rb
require 'active_support'

begin
  # This requires the full active_support.
  # ActiveSupport v3 and up are modular and 
  # need to be explicitly loaded.
  # Rescue in cases of ActiveSupport 2.3.2 and earlier.
  require 'active_support/all'
rescue
  # Do nothing, everything should be included
  # by default in older versions of ActiveSupport.
end

require 'action_mailer'

require File.join(File.dirname(__FILE__), *%w[cap_gun presenter])
require File.join(File.dirname(__FILE__), *%w[.. vendor action_mailer_tls lib smtp_tls])

# Tell everyone about your releases!  Send email notification after Capistrano deployments!  Rule the world!
# 
# We include the ActionMailer hack to play nice with Gmail, so that's a super easy way 
# to do this without setting up your own MTA.
#
# Example:
#
# Want to just shoot everyone an email about the latest status?
#
#   cap cap_gun:email
# 
# Include comments?
#
#   cap -s comment="hi mom" cap_gun:email
#
# Enable emails after every deploy by adding this to your deploy.rb:
# 
#   after "deploy", "cap_gun:email"
#
# Now, next time you deploy, you can optionally include comments:
#
#   cap -s comment="fix for bug #303" deploy
#
# See README for full install/config instructions.
module CapGun
  VERSION = '0.2.4'

  # This mailer is configured with a capistrano variable called "cap_gun_email_envelope"
  class Mailer < ActionMailer::Base

      def self.load_mailer_config(cap)
       raise ArgumentError, "You must define ActionMailer settings in 'cap_gun_action_mailer_config'" unless cap.cap_gun_action_mailer_config
       raise ArgumentError, "Need at least one recipient." if !cap.exists?(:cap_gun_email_envelope) || cap[:cap_gun_email_envelope][:recipients].blank?

       ActionMailer::Base.smtp_settings = cap.cap_gun_action_mailer_config
      end
      
      # Grab the options for emailing from capistrano[:cap_gun_email_envelope] (should be set in your deploy file)
      #
      # Valid options:
      #     :recipients     (required) an array or string of email address(es) that should get notifications
      #     :from           the sender of the notification, defaults to cap_gun@example.com
      #     :email_prefix   subject prefix, defaults to [DEPLOY]
      def deployment_notification(capistrano)
        presenter = Presenter.new(capistrano)
        
        content_type "text/plain"
        from         presenter.from
        recipients   presenter.recipients
        subject      presenter.subject
        body         presenter.body
      end
    end
    
end

if Object.const_defined?("Capistrano")

  Capistrano::Configuration.instance(:must_exist).load do

    namespace :cap_gun do
      desc "Send notification of the current release and the previous release via email."
      task :email, :roles => :app do
        CapGun::Mailer.load_mailer_config(self)
        if CapGun::Mailer.respond_to?(:deliver_deployment_notification)
          CapGun::Mailer.deliver_deployment_notification(self)
        else
          CapGun::Mailer.deployment_notification(self).deliver
        end
      end
    end
    
  end
end


panel/search_index.js
var search_data = {"index":{"searchIndex":["capgun","mailer","presenter","body()","branch()","comment()","convert_from_utc()","current_user()","deployed_to()","deployment_notification()","email_prefix()","exit_code()","from()","humanize_release_time()","load_mailer_config()","local_datetime_zone_offset()","local_timezone()","new()","previous_release_time()","recipients()","release_time()","scm_details()","scm_log()","scm_log_messages()","subject()","summary()","license","readme.markdown","cap_gun.rb","presenter.rb"],"info":[["CapGun","lib/cap_gun.rb","classes/CapGun.html"," < ","Tell everyone about your releases!  Send email notification after Capistrano deployments!  Rule the world!",1],["Mailer","CapGun","classes/CapGun/Mailer.html"," < ActionMailer::Base","This mailer is configured with a capistrano variable called \"cap_gun_email_envelope\" ",1],["Presenter","CapGun","classes/CapGun/Presenter.html"," < Object","",1],["body","CapGun::Presenter","classes/CapGun/Presenter.html#M000022","()","",2],["branch","CapGun::Presenter","classes/CapGun/Presenter.html#M000008","()","",2],["comment","CapGun::Presenter","classes/CapGun/Presenter.html#M000021","()","",2],["convert_from_utc","CapGun::Presenter","classes/CapGun/Presenter.html#M000015","(timestamp)","Use some DateTime magicrey to convert UTC to the current time zone When the whole world is on Rails 2.1",2],["current_user","CapGun::Presenter","classes/CapGun/Presenter.html#M000004","()","",2],["deployed_to","CapGun::Presenter","classes/CapGun/Presenter.html#M000007","()","",2],["deployment_notification","CapGun::Mailer","classes/CapGun/Mailer.html#M000010","(capistrano)","Grab the options for emailing from capistrano[:cap_gun_email_envelope] (should be set in your deploy",2],["email_prefix","CapGun::Presenter","classes/CapGun/Presenter.html#M000002","()","",2],["exit_code","CapGun::Presenter","classes/CapGun/Presenter.html#M000013","()","",2],["from","CapGun::Presenter","classes/CapGun/Presenter.html#M000003","()","",2],["humanize_release_time","CapGun::Presenter","classes/CapGun/Presenter.html#M000014","(path)","Gives you a prettier date/time for output from the standard Capistrano timestamped release directory.",2],["load_mailer_config","CapGun::Mailer","classes/CapGun/Mailer.html#M000005","(cap)","",2],["local_datetime_zone_offset","CapGun::Presenter","classes/CapGun/Presenter.html#M000016","()","",2],["local_timezone","CapGun::Presenter","classes/CapGun/Presenter.html#M000017","()","",2],["new","CapGun::Presenter","classes/CapGun/Presenter.html#M000000","(capistrano)","",2],["previous_release_time","CapGun::Presenter","classes/CapGun/Presenter.html#M000019","()","",2],["recipients","CapGun::Presenter","classes/CapGun/Presenter.html#M000001","()","",2],["release_time","CapGun::Presenter","classes/CapGun/Presenter.html#M000018","()","",2],["scm_details","CapGun::Presenter","classes/CapGun/Presenter.html#M000009","()","",2],["scm_log","CapGun::Presenter","classes/CapGun/Presenter.html#M000011","()","",2],["scm_log_messages","CapGun::Presenter","classes/CapGun/Presenter.html#M000012","()","",2],["subject","CapGun::Presenter","classes/CapGun/Presenter.html#M000020","()","",2],["summary","CapGun::Presenter","classes/CapGun/Presenter.html#M000006","()","",2],["LICENSE","files/LICENSE.html","files/LICENSE.html","","Copyright (c) 2009 Relevance, Inc.  Permission is hereby granted, free of charge, to any person obtaining",3],["README.markdown","files/README_markdown.html","files/README_markdown.html","","# CapGun    ## DESCRIPTION  Tell everyone about your releases!  Send email notification after Capistrano",3],["cap_gun.rb","files/lib/cap_gun_rb.html","files/lib/cap_gun_rb.html","","",3],["presenter.rb","files/lib/cap_gun/presenter_rb.html","files/lib/cap_gun/presenter_rb.html","","",3]],"longSearchIndex":["lib/cap_gun.rb","capgun","capgun","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::mailer","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::mailer","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","capgun::presenter","files/license.html","files/readme_markdown.html","files/lib/cap_gun_rb.html","files/lib/cap_gun/presenter_rb.html"]}}

lib/cap_gun/presenter.rb
require 'etc'

module CapGun
  class Presenter
    DEFAULT_SENDER = %("CapGun" <cap_gun@example.com>)
    DEFAULT_EMAIL_PREFIX = "[DEPLOY]"
    
    attr_accessor :capistrano
    
    def initialize(capistrano)
      self.capistrano = capistrano
    end

    def recipients
      capistrano[:cap_gun_email_envelope][:recipients]
    end

    def email_prefix
      capistrano[:cap_gun_email_envelope][:email_prefix] || DEFAULT_EMAIL_PREFIX
    end

    def from
      capistrano[:cap_gun_email_envelope][:from] || DEFAULT_SENDER
    end
    
    def current_user
      Etc.getlogin
    end

    def summary
      %[#{capistrano[:application]} was #{deployed_to} by #{current_user} at #{release_time}.]
    end

    def deployed_to
      return "deployed to #{capistrano[:rails_env]}" if capistrano[:rails_env]
      "deployed"
    end

    def branch
      "Branch: #{capistrano[:branch]}" unless capistrano[:branch].nil? || capistrano[:branch].empty?
    end

    def scm_details
      return unless [:git,:subversion].include? capistrano[:scm].to_sym
      <<-EOL
#{branch}
#{scm_log}
      EOL
      rescue
        nil
    end

    def scm_log
      "\nCommits since last release\n====================\n#{scm_log_messages}"
    end

    def scm_log_messages
      messages = case capistrano[:scm].to_sym
        when :git
          `git log #{previous_revision}..#{capistrano[:current_revision]} --pretty=format:%h:%s`
        when :subversion
          `svn log -r #{previous_revision.to_i+1}:#{capistrano[:current_revision]}`
        else
          "N/A"
      end
      exit_code.success? ? messages : "N/A"
    end

    def exit_code
      $?
    end

    # Gives you a prettier date/time for output from the standard Capistrano timestamped release directory.
    # This assumes Capistrano uses UTC for its date/timestamped directories, and converts to the local
    # machine timezone.
    def humanize_release_time(path)
      return unless path
      match = path.match(/(\d+)$/)
      return unless match
      local = convert_from_utc(match[1])
      local.strftime("%B #{local.day.ordinalize}, %Y %l:%M %p #{local_timezone}").gsub(/\s+/, ' ').strip
    end
    
    # Use some DateTime magicrey to convert UTC to the current time zone
    # When the whole world is on Rails 2.1 (and therefore new ActiveSupport) we can use the magic timezone support there.
    def convert_from_utc(timestamp)
      # we know Capistrano release timestamps are UTC, but Ruby doesn't, so make it explicit
      utc_time = timestamp << "UTC" 
      datetime = DateTime.parse(utc_time)
      datetime.new_offset(local_datetime_zone_offset)
    end
    
    def local_datetime_zone_offset
      @local_datetime_zone_offset ||= DateTime.now.offset
    end
    
    def local_timezone
      @current_timezone ||= Time.now.zone
    end
    
    def release_time
      humanize_release_time(capistrano[:current_release])
    end
    
    def previous_revision
      capistrano.fetch(:previous_revision, "n/a")
    end
    
    def previous_release_time 
      humanize_release_time(capistrano[:previous_release])
    end

    def subject
      "#{email_prefix} #{capistrano[:application]} #{deployed_to}"
    end
    
    def comment
      "Comment: #{capistrano[:comment]}.\n" if capistrano[:comment]
    end

    def body
<<-EOL
#{summary}
#{comment}
Deployment details
==================
Release: #{capistrano[:current_release]}
Release Time: #{release_time}
Release Revision: #{capistrano[:current_revision]}

Previous Release: #{capistrano[:previous_release]}
Previous Release Time: #{previous_release_time}
Previous Release Revision: #{previous_revision}

Repository: #{capistrano[:repository]}
Deploy path: #{capistrano[:deploy_to]}
Domain: #{capistrano[:domain]}
#{scm_details}
EOL
    end

  end
end


examples/presenter_example.rb
require File.join(File.dirname(__FILE__), *%w[example_helper])

describe CapGun::Presenter do 

  describe "scm details" do
  
    it "does nothing if the scm is not git nor subversion" do
      presenter = CapGun::Presenter.new(stub("capistrano", :scm => :mercurial))
      presenter.scm_details.should == nil
    end
  
    it "collects branch and git log details if git is scm" do
      capistrano = {:scm => :git, :branch => "edge"}
      presenter = CapGun::Presenter.new(capistrano)
      presenter.expects(:scm_log).returns("xxxxx")
    
      details = presenter.scm_details
      details.should include("Branch: edge")
      details.should include("xxxxx")
    end
  end

  describe "scm log messages" do
    describe "with git" do
      it "returns N/A if the git log call returns an error" do
        capistrano = {:scm => :git, :previous_revision => "previous-sha", :current_revision => "current-sha" }
        presenter = CapGun::Presenter.new(capistrano)
        presenter.stubs(:`).returns("fatal: Not a git repo")
        presenter.stubs(:exit_code).returns(stub("process status", :success? => false))
        presenter.scm_log_messages.should == "N/A"
      end

      it "calls git log with previous_rev..current_rev" do
        capistrano = {:scm => :git, :previous_revision => "previous-sha", :current_revision => "current-sha" }
        presenter = CapGun::Presenter.new(capistrano)
        presenter.stubs(:exit_code).returns(stub("process status", :success? => true))
        presenter.expects(:`).with(includes("git log previous-sha..current-sha"))
        presenter.scm_log_messages
      end
    end
    describe "with subversion" do
      it "returns N/A if the svn log call returns an error" do
        capistrano = {:scm => :subversion, :previous_revision => "42", :current_revision => "46" }
        presenter = CapGun::Presenter.new(capistrano)
        presenter.stubs(:`).returns("fatal: Not a working copy")
        presenter.stubs(:exit_code).returns(stub("process status", :success? => false))
        presenter.scm_log_messages.should == "N/A"
      end

      it "calls svn log with -r previous_rev+1:current_rev" do
        capistrano = {:scm => :subversion, :previous_revision => "42", :current_revision => "46" }
        presenter = CapGun::Presenter.new(capistrano)
        presenter.stubs(:exit_code).returns(stub("process status", :success? => true))
        presenter.expects(:`).with(includes("svn log -r 43:46"))
        presenter.scm_log_messages
      end
    end
  end

  describe "release time" do
  
    before do # make DateTime act as if local timezone is CDT
      @presenter = CapGun::Presenter.new(nil)
      @presenter.stubs(:local_timezone).returns("CDT")
      @presenter.stubs(:local_datetime_zone_offset).returns(Rational(-1,6))
    end
  
    it "returns nil for weird release path" do
      @presenter.humanize_release_time("/data/foo/my_release").should == nil
    end
  
    it "parse datetime from release path" do
      @presenter.humanize_release_time("/data/foo/releases/20080227120000").should == "February 27th, 2008 8:00 AM CDT"
    end
  
    it "converts time from release into localtime" do
      @presenter.humanize_release_time("/data/foo/releases/20080410040000").should == "April 10th, 2008 12:00 AM CDT"
    end
  
  end

  describe "from and to emails" do
    
    it "gets recipients from email envelope" do
      capistrano = { :cap_gun_email_envelope => { :recipients => ["foo@here.com", "bar@here.com"] } }
      presenter = CapGun::Presenter.new(capistrano)
      presenter.recipients.should == ["foo@here.com", "bar@here.com"]
    end

    it "should have a default sender" do
      capistrano = { :cap_gun_email_envelope => { } }
      presenter = CapGun::Presenter.new(capistrano)
      presenter.from.should == "\"CapGun\" <cap_gun@example.com>"
    end
  
    it "should override sender from email envelope" do
      capistrano = { :cap_gun_email_envelope => { :from => "booyakka!@example.com" } }
      presenter = CapGun::Presenter.new(capistrano)
      presenter.from.should == "booyakka!@example.com"
    end
    
  end
  
  describe "previous_revision" do
    
    it "returns n/a if not set" do
      CapGun::Presenter.new({}).previous_revision.should == "n/a"
    end
    
    it "returns the previous revision if set" do
      CapGun::Presenter.new({:previous_revision => "100"}).previous_revision.should == "100"
    end
    
  end

end

examples/cap_gun_example.rb
require File.join(File.dirname(__FILE__), *%w[example_helper])

describe CapGun do

  it "uses action mailer hack" do
    Net::SMTP.new('').respond_to?(:starttls, true).should == true
  end

  describe "mail settings" do
    
    it "raises if we don't have settings" do
      capistrano = stub_everything
      lambda { 
        CapGun::Mailer.load_mailer_config(capistrano) 
      }.should raise_error(ArgumentError, "You must define ActionMailer settings in 'cap_gun_action_mailer_config'")
    end
  
    it "gets action mailer config from capistrano" do
      capistrano = stub(:cap_gun_action_mailer_config => {:account => "foo@gmail.com", :password => "password"}, :exists? => true)
      capistrano.stubs(:[]).returns({:recipients => 'foo'})
      CapGun::Mailer.load_mailer_config(capistrano)
      ActionMailer::Base.smtp_settings.should == {:account => "foo@gmail.com", :password => "password"}
    end
    
    it "raises if don't have a cap gun email envelope" do
      capistrano = stub_everything(:cap_gun_action_mailer_config => {}, :exists? => false)
      lambda { 
        CapGun::Mailer.load_mailer_config capistrano 
      }.should raise_error(ArgumentError)
    end
    
    it "raises if we don't have at least one recipient" do
      capistrano = stub_everything(:cap_gun_action_mailer_config => {}, :cap_gun_email_envelope => {})
      lambda { 
        CapGun::Mailer.load_mailer_config capistrano 
      }.should raise_error(ArgumentError)
      capistrano = stub_everything(:cap_gun_action_mailer_config => {}, :cap_gun_email_envelope => {:recipients => []})
      lambda { 
        CapGun::Mailer.load_mailer_config capistrano 
      }.should raise_error(ArgumentError)
    end
    
  end
  
  describe CapGun::Mailer do
    
    it "calls Net::SMTP to send the mail correctly (we test this because SMTP internals changed between 1.8.6 and newer versions of Ruby)" do
      ActionMailer::Base.smtp_settings = {
        :address => "smtp.gmail.com",
        :port => 587,
        :domain => "foo.com",
        :authentication => :plain,
        :user_name => "username",
        :password => "password"
      }
    
      capistrano = { :current_release => "/data/foo", :previous_release => "/data/foo", :cap_gun_email_envelope => {:recipients => ["joe@example.com"]} }        
      smtp = Net::SMTP.new('gmail.com', 25)
      Net::SMTP.expects(:new).returns(smtp)
      smtp.expects(:start)
      CapGun::Mailer.deliver_deployment_notification capistrano
    end
    
  end
  
end